central/role/datastore/datastore_impl.go

```diff
@@ -32,6 +32,117 @@ type dataStoreImpl struct {
 	lock sync.RWMutex
 }
 
+func (ds *dataStoreImpl) UpsertRole(ctx context.Context, newRole *storage.Role) error {
+	if err := sac.VerifyAuthzOK(roleSAC.WriteAllowed(ctx)); err != nil {
+		return err
+	}
+	if err := rolePkg.ValidateRole(newRole); err != nil {
+		return errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	oldRole, exists, err := ds.roleStorage.Get(ctx, newRole.GetName())
+	if err != nil {
+		return err
+	}
+	if exists {
+		if err := verifyRoleOriginMatches(ctx, oldRole); err != nil {
+			return err
+		}
+	}
+	if err := verifyRoleOriginMatches(ctx, newRole); err != nil {
+		return err
+	}
+
+	permissionSet, accessScope, err := ds.verifyRoleReferencesExist(ctx, newRole)
+	if err != nil {
+		return err
+	}
+	if err := verifyPermissionSetOriginMatches(ctx, permissionSet); err != nil {
+		return err
+	}
+	if err := verifyAccessScopeOriginMatches(ctx, accessScope); err != nil {
+		return err
+	}
+
+	// Constraints ok, write the object. We expect the underlying store to
+	// verify there is no role with the same name.
+	if err := ds.roleStorage.Upsert(ctx, newRole); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ds *dataStoreImpl) UpsertPermissionSet(ctx context.Context, newPS *storage.PermissionSet) error {
+	if err := sac.VerifyAuthzOK(roleSAC.WriteAllowed(ctx)); err != nil {
+		return err
+	}
+	if err := rolePkg.ValidatePermissionSet(newPS); err != nil {
+		return errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	oldPS, exists, err := ds.permissionSetStorage.Get(ctx, newPS.GetId())
+	if err != nil {
+		return err
+	}
+	if exists {
+		if err := verifyPermissionSetOriginMatches(ctx, oldPS); err != nil {
+			return err
+		}
+	}
+	if err := verifyPermissionSetOriginMatches(ctx, newPS); err != nil {
+		return err
+	}
+
+	// Constraints ok, write the object. We expect the underlying store to
+	// verify there is no permission set with the same name.
+	if err := ds.permissionSetStorage.Upsert(ctx, newPS); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ds *dataStoreImpl) UpsertAccessScope(ctx context.Context, newScope *storage.SimpleAccessScope) error {
+	if err := sac.VerifyAuthzOK(roleSAC.WriteAllowed(ctx)); err != nil {
+		return err
+	}
+	if err := rolePkg.ValidateSimpleAccessScope(newScope); err != nil {
+		return errors.Wrap(errox.InvalidArgs, err.Error())
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	oldScope, exists, err := ds.accessScopeStorage.Get(ctx, newScope.GetId())
+	if err != nil {
+		return err
+	}
+	if exists {
+		if err := verifyAccessScopeOriginMatches(ctx, oldScope); err != nil {
+			return err
+		}
+		if err := verifyAccessScopeOriginMatches(ctx, newScope); err != nil {
+			return err
+		}
```

Comment:
We should check the access scope's origin also when its a new one - which we currently wouldn't do.

Reply:
That's a very bad Ctrl-C Ctrl-V 😞